### PR TITLE
functions for decimal(p,s)

### PR DESCRIPTION
**Pull request by ilfrin**
This set of commits introduces functions (aggregate and scalar) for decimal(p,s).

The implementation for aggregate function uses the "old" way (overriding specialize()). Once the "new way", the annotation framework for aggregate functions, accommodates parametrised types this should be refactored to leverage that.

The scalar functions are implemented using the builder approach.

Most implementations use BigDecimal and BigInteger freely. Once there is a fast LongDecimalType implementation we can look to those and improve their performance. At this point I was going for correctness.

### CODE REVIEW

**Comment by ilfrin** [comment]
@losipiuk this PR supersedes https://github.com/Teradata/presto/pull/115 please review.

**Comment by ilfrin** [comment]
working on travis, please review

**Comment by ilfrin** [comment]
travis should be fine now

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 1]
```diff
@@ -291,6 +430,146 @@ public static double round(@SqlType(StandardTypes.DOUBLE) double num, @SqlType(S
...
+        if (inputScale == 0) {
+            return unscaledValueToSlice(num);
+        }
+        if (num.signum() < 0) {
+            return unscaledValueToSlice(roundLongDecimal(num.negate(), rescaleFactor).negate());
+        }
+        return unscaledValueToSlice(roundLongDecimal(num, rescaleFactor));
+    }
+
+    public static BigInteger roundLongDecimal(BigInteger num, BigInteger rescaleFactor)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: why public?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 2]
```diff
@@ -291,6 +430,146 @@ public static double round(@SqlType(StandardTypes.DOUBLE) double num, @SqlType(S
...
+    }
+
+    private static SqlScalarFunction decimalRoundNFunction()
+    {
+        Signature signature = Signature.builder()
+                .kind(SCALAR)
+                .name("round")
+                .literalParameters("num_precision", "num_scale", "result_precision")
+                .longVariableConstraints(
+                        longVariableCalculation("result_precision", "min(38, num_precision + (num_scale - max(num_scale-1, 0)))"))
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: comment this

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java [hunk 3]
```diff
@@ -14,12 +14,15 @@
 package com.facebook.presto.operator.scalar;
 
 import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.type.SqlDecimal;
```
Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java:17]: add tests for overflow due to `round`, `roundN`, `ceil`, and `floor`

Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java:17]: only roundN on the second though - other cannot overflow - right?

**Comment by losipiuk** [comment]
looks good

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 1 above)
Comment (ilfrin) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: made private

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 2 above)
Comment (ilfrin) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: done

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java (hunk 3 above)
Comment (ilfrin) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java:17]: right, only roundN can overflow

Comment (ilfrin) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java:17]: added test for overflow

**Comment by ilfrin** [comment]
applied comments

**Comment by ilfrin** [comment]
closing, will keep this as a separate branch, renamed branch to https://github.com/Teradata/presto/tree/decimal-functions
